src/timesync/timesyncd-manager.h

```diff
@@ -83,7 +83,6 @@ struct Manager {
 
         /* last change */
         bool jumped;
-        bool sync;
```

Comment:
hmm, I wonder if just dropping this test altogether is the right approach. I think we should probably tweak things a bit, but stick to the general logic that we need a sync first before we persist the time to disk.

Why? The thing is that we got reports in the past of systems that on very first boot up with basically random time in the battery buffered RTC, which then creates all kinds of issues if we persist that to disk, because the general assumption is that time progresses but never goes backwards.

I'd thus like to tweak the logic here, so that we persist the current time once:

1. a synchronization with NTP is reached at least once
2. the time was previously synced to disk already, i.e. access(CLOCK_FILE, F_OK) succeeds

That should deliver what you are looking for, but still cover the cases with the randomized time as discussed above relatively nicely.

Reply:
The only problem is that offline systems (the primary target for these patches) will never get an NTP sync and won't ever benefit from this behavior. :smile: 

What about this idea:

We keep `bool sync;` and I will set it to `true` inside `manager_save_time_and_rearm()`, so it will be set under one of two conditions: (1) NTP sync or (2) SaveIntervalSec has expired.

If we keep SaveIntervalSec option off by default (which I propose we do), this doesn't change the current behavior. And for offline systems with this option set, they will need to be up for at least SaveIntervalSec for the time to be persisted.

Reply:
> I think we should probably tweak things a bit, but stick to the general logic that we need a sync first before we persist the time to disk.

I can see that arguments can be made along these lines, but to add other angles too, it would be a little unfortunate from the perspective of distros like Endless and Raspberry Pi OS, which contemplate never-online users without RTC batteries. In such environment, timesyncd can replace [fake-hwclock](https://manpages.debian.org/jessie/fake-hwclock/fake-hwclock.8.en.html) if the behaviour is tweaked as proposed here. You could argue that its out of scope of timesyncd, but if we do not do so, then fake-hwclock is still required for the offline case and that may present some awkward interactions between two different services that both save time on shutdown and restore time on startup with different conditions.

> Why? The thing is that we got reports in the past of systems that on very first boot up with basically random time in the battery buffered RTC, which then creates all kinds of issues if we persist that to disk, because the general assumption is that time progresses but never goes backwards.

I don't quite follow this example though. OK, you got a random time on first boot, but what are the scenarios we can face now?

1. System goes online, timesyncd gets synchronized from ntp and has its existing behaviour of writing the time to disk upon synchronization and upon shutdown. We're all set for next boot, time will move forwards.
2. The system does not go online but the user swiftly corrects the system time manually, and also writes this to the battery-backed RTC.
    * If timesyncd retains the existing behaviour of not doing anything, the fixed time in the RTC will continue to increment while the system is off, and on next boot, it will be accurate, having elapsed while the system was off. Time is still moving forwards.
    * If timesyncd is tweaked to save the time on shutdown, it doesn't change anything. On next boot, the RTC has the correct time, having elapsed while the system was off. The time saved to disk will be ignored since it is in the past compared to the current RTC reading. Time is still moving forwards.
3. The system does not go online and does not have the time corrected by the user.
    * If timesyncd retains the existing behaviour of not doing anything, the random time in the RTC will continue to increment while the system is off, and on next boot, it will be the original random time plus the time elapsed. Time is still moving forwards, even though the clock is totally wrong.
    * If timesyncd is tweaked to save the time on shutdown, it doesn't change anything. On next boot, the RTC reads the original random time plus the time elapsed since then. The time saved to disk will be ignored since it is in the past compared to the current RTC reading. Time is still moving forwards.

Unless I'm missing something, having timesyncd save the time on shutdown doesn't cause complications for the random battery backed RTC case.

> I'd thus like to tweak the logic here, so that we persist the current time once:
> 1. a synchronization with NTP is reached at least once
> 2. the time was previously synced to disk already, i.e. access(CLOCK_FILE, F_OK) succeeds

Are these AND- or OR- conditions?
Note that the 2nd point is generally always true since timesyncd creates the file on startup if it does not exist (stamping with systemd build date).

Reply:
fully offline operation definitely makes sense to support. I am just looking for a way we can make this stuff "just work", i.e. a way that works automatically the right way on all systems, i.e. all of these cases:

1. fully offline systems without RTC
2. fully offline systems with an RTC but no battery backing (which is a thing you say, though I don't get what the point is, given that the system clock should do everything the RTC does anyway.)
3. fully offline systems with a working, battery backed RTC
4. systems which come up with a random time in the RTC
5. systems where we get an NTP sync

if we can cover these three cases in a way that no further config is necessary this would be ideal in my eyes.

Reply:
That's good to know. I'm still not clear on exactly why this approach does not satisfy those cases though.

Seems like you were mostly worried about number 4? But I don't really understand - if the user corrects the time then we will want to save that (regardless of access to connectivity). If the user doesn't correct the time and the time is random, we're going to want to save that also, to ensure that time at least flows forward.

Reply:
Answered below.